x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/action/TransportUpdateDataFrameTransformAction.java

```diff
@@ -0,0 +1,232 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.dataframe.action.TransportPutDataFrameTransformAction.buildPrivilegeCheck;
+
+public class TransportUpdateDataFrameTransformAction extends TransportMasterNodeAction<Request, Response> {
+
+    private final XPackLicenseState licenseState;
+    private final Client client;
+    private final DataFrameTransformsConfigManager dataFrameTransformsConfigManager;
+    private final SecurityContext securityContext;
+    private final DataFrameAuditor auditor;
+
+    @Inject
+    public TransportUpdateDataFrameTransformAction(Settings settings, TransportService transportService, ThreadPool threadPool,
+                                                   ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                                   ClusterService clusterService, XPackLicenseState licenseState,
+                                                   DataFrameTransformsConfigManager dataFrameTransformsConfigManager, Client client,
+                                                   DataFrameAuditor auditor) {
+        super(UpdateDataFrameTransformAction.NAME, transportService, clusterService, threadPool, actionFilters,
+                Request::new, indexNameExpressionResolver);
+        this.licenseState = licenseState;
+        this.client = client;
+        this.dataFrameTransformsConfigManager = dataFrameTransformsConfigManager;
+        this.securityContext = XPackSettings.SECURITY_ENABLED.get(settings) ?
+            new SecurityContext(settings, threadPool.getThreadContext()) : null;
+        this.auditor = auditor;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected Response read(StreamInput in) throws IOException {
+        return new Response(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState clusterState, ActionListener<Response> listener) {
+
+        if (!licenseState.isDataFrameAllowed()) {
+            listener.onFailure(LicenseUtils.newComplianceException(XPackField.DATA_FRAME));
+            return;
+        }
+
+        XPackPlugin.checkReadyForXPackCustomMetadata(clusterState);
+
+        // set headers to run data frame transform as calling user
+        Map<String, String> filteredHeaders = threadPool.getThreadContext().getHeaders().entrySet().stream()
+                    .filter(e -> ClientHelper.SECURITY_HEADER_FILTERS.contains(e.getKey()))
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        DataFrameTransformConfigUpdate update = request.getUpdate();
+        update.setHeaders(filteredHeaders);
+
+        String transformId = request.getId();
+
+        // GET transform and attempt to update
+        dataFrameTransformsConfigManager.getTransformConfiguration(request.getId(), ActionListener.wrap(
```

Comment:
ideally update should use optimistic locking, which means you save the doc version and send it together with the index request for the update